compiler/parser/parser.cpp

```diff
@@ -505,14 +505,44 @@ static ModuleSymbol* parseMod(const char* modName, bool isInternal) {
 static bool containsOnlyModules(BlockStmt* block, const char* path);
 static void addModuleToDoneList(ModuleSymbol* module);
 
+//
+// This is a check to see whether we've already parsed this file
+// before to avoid re-parsing the same thing twice which can result in
+// defining its modules twice.
+//
+static bool haveAlreadyParsed(const char* path) {
+  static std::set<std::string> parsedPaths;
+
+  // normalize the path if possible via realpath() and use 'path' otherwise
+  const char* normpath = chplRealPath(path);
+  if (normpath == NULL) {
+    normpath = path;
+  }
+
+  // check whether we've seen this path before
+  if (parsedPaths.count(normpath) > 0) {
+    // if so, indicate it
+    return true;
+  } else {
+    // otherwise, add it to our set and list of paths
+    parsedPaths.insert(normpath);
+    gFilenameLookup.push_back(path);
+    return false;
+  }
+}
+
+
 static ModuleSymbol* parseFile(const char* path,
                                ModTag      modTag,
                                bool        namedOnCommandLine) {
   ModuleSymbol* retval = NULL;
 
+  // Make sure we haven't already parsed this file
+  if (haveAlreadyParsed(path)) {
+    return NULL;
+  }
 
   if (FILE* fp = openInputFile(path)) {
-    gFilenameLookup.push_back(path);
```

Comment:
I think it's confusing that gFilenameLookup is modified in `haveAlreadyParsed`. Could we just do the push_back here as we did before?

Reply:
Yeah, that's a good idea.  In an intermediate version I was using gFilenameLookup for both the set and its traditional use, so it made sense to both check for updates and modify it in the same place.  Now that I've split that back out into its own set datatype (rationale: because realPaths are absolute whereas these gFilenameLookup paths are often relative and more digestible when they are), I think you're right that I should put this back where it was as well.